numpy/core/src/umath/clip.cpp

```diff
@@ -0,0 +1,152 @@
+/**
+ * This module provides the inner loops for the clip ufunc
+ */
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "Python.h"
+
+#include "numpy/halffloat.h"
+#include "numpy/npy_math.h"
+#include "numpy/ndarraytypes.h"
+#include "numpy/npy_common.h"
+#include "numpy/utils.h"
+#include "fast_loop_macros.h"
+
+namespace npy {
+
+  struct integral_tag {};
+  struct floating_point_tag {};
+  struct complex_tag {};
+  struct date_tag {};
+
+  struct bool_tag : integral_tag { using type = npy_bool; };
+  struct byte_tag : integral_tag {using type = npy_byte; } ;
+  struct ubyte_tag : integral_tag {using type = npy_ubyte; } ;
+  struct short_tag : integral_tag {using type = npy_short; } ;
+  struct ushort_tag : integral_tag {using type = npy_ushort; } ;
+  struct int_tag : integral_tag {using type = npy_int; } ;
+  struct uint_tag : integral_tag {using type = npy_uint; } ;
+  struct long_tag : integral_tag {using type = npy_long ; } ;
+  struct ulong_tag : integral_tag {using type = npy_ulong ; } ;
+  struct longlong_tag : integral_tag {using type = npy_longlong ; } ;
+  struct ulonglong_tag : integral_tag {using type = npy_ulonglong ; } ;
+  struct half_tag {using type = npy_half ; } ;
+  struct float_tag : floating_point_tag {using type = npy_float ; } ;
+  struct double_tag : floating_point_tag {using type = npy_double ; } ;
+  struct longdouble_tag : floating_point_tag {using type = npy_longdouble ; } ;
+  struct cfloat_tag : complex_tag {using type = npy_cfloat ; } ;
+  struct cdouble_tag : complex_tag {using type = npy_cdouble ; } ;
+  struct clongdouble_tag : complex_tag {using type = npy_clongdouble ; } ;
+  struct datetime_tag : date_tag {using type = npy_datetime ; } ;
+  struct timedelta_tag : date_tag {using type = npy_timedelta ; } ;
+
+}
+
+template<class T>
+T _NPY_MIN(T a, T b, npy::integral_tag const &) { return PyArray_MIN(a, b); }
+template<class T>
+T _NPY_MAX(T a, T b, npy::integral_tag const &) { return PyArray_MAX(a, b); }
+
+template<class T>
+T _NPY_MIN(T a, T b, npy::half_tag const &) { return npy_half_isnan(a) || npy_half_le(a, b) ? (a) : (b); }
+template<class T>
+T _NPY_MAX(T a, T b, npy::half_tag const &) { return npy_half_isnan(a) || npy_half_ge(a, b) ? (a) : (b); }
```

Comment:
Why is this a template, instead of just
```cpp
npy_half _NPY_MAX(npy_half a, npy_half b, floating_point_tag const &)
```

This would remove the need to have per-dtype tag types.

Reply:
per-dtype tag types are useful because some dtype alias to the same type while having different meaning (`npy_long` and `npy_datetime` if I recall correctly) so we can't relyon the actual dtype for the dispatch.

Reply:
Okay, that's something worth fixing IMO. Overloading is one of the main benefits to C++. Tag issues notwithstanding,  this still doesn't need to be a template:
```cpp
npy_half _NPY_MAX(npy_half a, npy_half b, npy::half_tag const &)
```

Reply:
I would prefer using templates only where absolutely required.